### DISCUSSION TITLE
How should we monitor JS bundle size creep in CI?

### DISCUSSION THREAD

**Discussion by mistercrunch** [apache/superset#29803] [Ideas]
This recent series of PRs https://github.com/apache/superset/labels/dashboard%3Aperformance highlights the fact that our Javascript bundles have gotten out of control over the years. 

In retrospect, a simple `npm install {some_useful_library}` can have significant impact on perf, and we should setup some automated mechanism to report - on a per-PR basis - on how PRs are affecting bundle sizes / perf. We're also realizing that not only the `npm install` is important, but how we specify the actual `import` statements can have significant impact on bundle size. 

Also note that we now have ways to load bundles dynamically, meaning part of introducing new libraries might prompt for decisions around what should be in core bundle and what should be loaded dynamically. 

In any case, we need better automation/tooling around understanding bundle size. This discussion is about defining the right way to monitor/alert using CI in PRs that affect bundle size.

Some ideas:
- have a GHA workflow that triggers only when package-lock.json changes and produce a report on bundle size
- write a script or find some utility that produces a proper report on bundle size, here the visualizations (treemap) have proven pretty useful
- run the script for the before and after - checkout the "merge base" and run the report against that too
- do some sort of diff between the current PR commit and its merge-base
- fail the check if the bundle size has grown by anything significant (what's the right threshold here?)
- make the workflow/check not required, meaning the committer/merger can review the failed CI task, and make a call to merge despite the bundle size increase. Maybe they write a note as they merged "I reviewed the bundle size increase, and the +8KB is well worth it for the feature that this library provides", or "this adds 200KB to the bundles, but it is loaded dynamically only when the DatasetEditor is shown, ...."

**Comment by kgabryje** [comment]
Initially, I was thinking of using some bundle size analyser to detect a bundle size and fail if the size increased by more than X.
However, this approach won't detect increased network activity caused by importing a module that should be imported lazily rather then eagerly. Example from PR https://github.com/apache/superset/pull/29791 - if someone writes a statement `import moment from "moment-timezone"`, that won't affect the bundle size at all, but it will break lazy loading.

I wonder if we could utilise Cypress to open a dashboard page (or chart, sql lab, home...), intercept all requests and sum their sizes. If the size is larger than expected that could mean that either tree shaking of a large library is broken, or a large library is imported eagerly.
WDYT?

<img width="1020" alt="Screenshot 2024-08-01 at 15 29 32" src="https://github.com/user-attachments/assets/446a315a-52fb-4144-b0ca-0fb91d7b05ca">

**Comment by mistercrunch** [comment]
I know there are ways to write custom lint rules around imports. I believe @betodealmeida did some of that on the python side, something like "don't allow importing sqlparse because we use sqlglot now". I'm wondering if writing custom `eslint` rules to prevent certain kind of imports would be a good way to enforce the sort of imports you're trying to avoid?

Would be nice to avoid deep-cypress powered solutions if possible ...

**Comment by kgabryje** [comment]
We could also take a higher level approach and integrate Google Lighthouse in our CI, which would fail if some metrics (like LCP or speed index) fall by X. Looks like it might be easy to integrate with Github Actions https://github.com/GoogleChrome/lighthouse-ci
